crates/chain-config/src/config/codec.rs

```diff
@@ -0,0 +1,262 @@
+pub mod in_memory;
+pub mod json;
+pub mod parquet;
+
+use std::{
+    fmt::Debug,
+    path::{
+        Path,
+        PathBuf,
+    },
+};
+
+use ::parquet::basic::{
+    Compression,
+    GzipLevel,
+};
+
+use crate::{
+    CoinConfig,
+    ContractConfig,
+    MessageConfig,
+    StateConfig,
+};
+
+use super::{
+    contract_balance::ContractBalance,
+    contract_state::ContractState,
+};
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Group<T> {
+    pub index: usize,
+    pub data: Vec<T>,
+}
+
+type GroupResult<T> = anyhow::Result<Group<T>>;
+pub trait GroupDecoder<T>: Iterator<Item = GroupResult<T>> {}
+
+pub trait StateEncoder {
+    fn write_coins(&mut self, elements: Vec<CoinConfig>) -> anyhow::Result<()>;
+    fn write_contracts(&mut self, elements: Vec<ContractConfig>) -> anyhow::Result<()>;
+    fn write_messages(&mut self, elements: Vec<MessageConfig>) -> anyhow::Result<()>;
+    fn write_contract_state(
+        &mut self,
+        elements: Vec<ContractState>,
+    ) -> anyhow::Result<()>;
+    fn write_contract_balance(
+        &mut self,
+        elements: Vec<ContractBalance>,
+    ) -> anyhow::Result<()>;
+    fn close(self: Box<Self>) -> anyhow::Result<()>;
+}
+
+pub type DynGroupDecoder<T> = Box<dyn GroupDecoder<T>>;
```

Comment:
Does this need to be dynamic?

Reply:
What decoder actually gets used (json/parquet) is decided at runtime based on the file extension of the configs. So we resort to trait object polymorphism here to make higher level methods agnostic of the format.

Reply:
> So a dyn dispatch has to happen somewhere since we don't know the type until the app is ran.

Couldn't you use an enum:
```rs
#[non_exhaustive]
pub enum ConcreteDecoder<T> {
    JSON(JSONDecoder<T>),
    Parquet(ParquetDecoder<T>),
    InMemory(InMemoryDecoder<T>),
}

impl<T> GroupDecoder<T> for ConcreteDecoder<T> {
    ...
}
```

> There might be way to make it non-dynamic via an enum and lot's of if/match statements but it would affect proper separation and make it messier overall.

This feels like the idiomatic way to me. I have a general aversion to `dyn Trait` though :P Almost never use it.

Reply:
Decided your suggestion a try. Turned out great in the end IMO. Lemme know your thoughts when you get the time.

Reply:
Nice. On a quick glance I'm not sure I follow the use of the `Iterator`, it seems like it could be valuable to capture the behavior in domain terms, rather than implementation terms. But I'm gonna give it a read in the full context and I'm sure it'll make more sense--I don't think I have a full picture yet.